pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/properties/PreserveStashesJobProperty.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.pipeline.modeldefinition.properties;
+
+import hudson.Extension;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import hudson.util.FormValidation;
+import jenkins.model.OptionalJobProperty;
+import org.jenkinsci.Symbol;
+import org.jenkinsci.plugins.workflow.flow.StashManager;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class PreserveStashesJobProperty extends OptionalJobProperty<WorkflowJob> {
+
+    private static final Logger LOGGER = Logger.getLogger(PreserveStashesJobProperty.class.getName());
+
+    private int buildCount = 10;
```

Comment:
Strange default. 1 would make some sense, perhaps. Consider the normal case that you want to permit resumption of only the most recent build, and tune the default accordingly.

Reply:
Sounds reasonable.

Reply:
Sorry, I don't really have the context of the code or the details of this discussion, but from this related issue, https://issues.jenkins-ci.org/browse/JENKINS-45455...

*Use cases*
* Restarting a successful stage so to rollback or redeploy code

Sounds like this PR discussion is about limiting the restart to the most recent build. Seems like that's at odds with the above use case?

Reply:
@ljohnston You'd want to specify a higher number to keep in that case - we're just talking about setting the default value, not the max.

Reply:
Ah, got it... thx!

Reply:
Status?

Reply:
Still considering.

Reply:
Yeah, 1 makes sense for the default. Updating.